dialects/neo4j.go

```diff
@@ -265,23 +265,99 @@ func (n *Neo4jRuntime) Compile(cradle *e.QueryCradle) (string, error) {
 
-					className = reflect.TypeOf(cradle.Out).Elem().Name()
+					kvp["className"] = reflect.TypeOf(cradle.Out).Elem().Name()
 				} else if reflect.TypeOf(cradle.Out).Kind() == reflect.Struct {
-					className = reflect.TypeOf(cradle.Out).Name()
+					kvp["className"] = reflect.TypeOf(cradle.Out).Name()
+				}
+
+				genQuery = fmt.Sprintf("MATCH (%s: %s {%s}) DETACH DELETE %s", kvp["nodeName"], kvp["className"],
+					cypherA, kvp["nodeName"])
+			}
+
+			return genQuery, nil
+		case e.RelationX:
+			exp, err := cradle.Exps.Get()
+			if err != nil {
+				return "", err
+			}
+
+			if reflect.TypeOf(exp).Kind() == reflect.Ptr {
+				if reflect.TypeOf(exp).Elem().Kind() == reflect.Struct {
+					kvp["classNameX"] = reflect.TypeOf(exp).Elem().Name()
+				} else if reflect.TypeOf(exp).Elem().Kind() == reflect.Slice &&
+					reflect.TypeOf(exp).Elem().Elem().Kind() == reflect.Struct {
+					return "", e.Error(e.ExpectedStructNotSlice)
+				}
+			} else if reflect.TypeOf(exp).Kind() == reflect.Struct {
+				kvp["classNameX"] = reflect.TypeOf(exp).Name()
+			}
+
+			lucyExp := e.Marshal(exp)
+			e.SanitizeExp(lucyExp)
+
+			kvp["matchExpX"] = n.marshalToCypherExp(lucyExp)
+		case e.RelationY:
+			exp, err := cradle.Exps.Get()
+			if err != nil {
+				return "", err
+			}
+
+			if reflect.TypeOf(exp).Kind() == reflect.Ptr {
+				if reflect.TypeOf(exp).Elem().Kind() == reflect.Struct {
+					kvp["classNameY"] = reflect.TypeOf(exp).Elem().Name()
+				} else if reflect.TypeOf(exp).Elem().Kind() == reflect.Slice &&
+					reflect.TypeOf(exp).Elem().Elem().Kind() == reflect.Struct {
+					return "", e.Error(e.ExpectedStructNotSlice)
 				}
+			} else if reflect.TypeOf(exp).Kind() == reflect.Struct {
+				kvp["classNameY"] = reflect.TypeOf(exp).Name()
+			}
 
-				genQuery = fmt.Sprintf("MATCH (%s: %s {%s}) DETACH DELETE %s", nodeName, className, cypherA, nodeName)
+			lucyExp := e.Marshal(exp)
+			e.SanitizeExp(lucyExp)
+
+			kvp["matchExpY"] = n.marshalToCypherExp(lucyExp)
+		case e.By:
+			exp, err := cradle.Exps.Get()
+			if err != nil {
+				return "", err
 			}
 
+			relName := exp.(string)
+			genQuery := fmt.Sprintf("MATCH (n: %s {%s}), (m: %s {%s}) CREATE (n)-[:%s]->(m)", kvp["classNameX"],
+				kvp["matchExpX"], kvp["classNameY"], kvp["matchExpY"], relName)
 			return genQuery, nil
+		case e.MTRelation:
+			kvp["Context.Find"] = "Relation"
+
+			exp, err := cradle.Exps.Get()
+			if err != nil {
+				return "", err
+			}
+
+			pExp := exp.(e.Exp)
+
+			kvp["classNameX"] = e.GetTypeName(pExp["out"])
+
+			outExp := e.Marshal(pExp["out"])
+			e.SanitizeExp(outExp)
+			kvp["cypherA"] = n.marshalToCypherExp(outExp)
+			kvp["relName"] = pExp["relation"].(string)
+			break
 		}
 	}
 
 	return "", nil
 }
 
 func (n *Neo4jRuntime) Execute(query string, cradle *e.QueryCradle, target interface{}) error {
-	result, err := n.session.Run(query, map[string]interface{}{})
+	var result neo4j.Result
+	var err error
+	if n.isTransaction {
```

Comment:
Redundant. Provided we use pointers to complex types, we can simply do `if n.transaction != nil`